timeinterval/timeinterval.go

```diff
@@ -0,0 +1,473 @@
+			} else {
+				begin = validDates.Begin
+			}
+			if validDates.End < 0 {
+				end = daysInMonth + validDates.End + 1
+			} else {
+				end = validDates.End
+			}
+			// Skip clamping if the beginning date is after the end of the month
+			if begin > daysInMonth {
+				continue
+			}
+			// Clamp to the boundaries of the month to prevent crossing into other months
+			begin = clamp(begin, -1*daysInMonth, daysInMonth)
+			end = clamp(end, -1*daysInMonth, daysInMonth)
+			if t.Day() >= begin && t.Day() <= end {
+				in = true
+				break
+			}
+		}
+		if !in {
+			return false
+		}
+	}
+	if tp.Months != nil {
+		in := false
+		for _, validMonths := range tp.Months {
+			if t.Month() >= time.Month(validMonths.Begin) && t.Month() <= time.Month(validMonths.End) {
+				in = true
+				break
+			}
+		}
+		if !in {
+			return false
+		}
+	}
+	if tp.Weekdays != nil {
+		in := false
+		for _, validDays := range tp.Weekdays {
+			if t.Weekday() >= time.Weekday(validDays.Begin) && t.Weekday() <= time.Weekday(validDays.End) {
+				in = true
+				break
+			}
+		}
+		if !in {
+			return false
+		}
+	}
+	if tp.Years != nil {
+		in := false
+		for _, validYears := range tp.Years {
+			if t.Year() >= validYears.Begin && t.Year() <= validYears.End {
+				in = true
+				break
+			}
+		}
+		if !in {
+			return false
+		}
+	}
+	return true
+}
+
+// Converts a string of the form "HH:MM" into the number of minutes elapsed in the day
+func parseTime(in string) (mins int, err error) {
+	if !validTimeRE.MatchString(in) {
+		return 0, fmt.Errorf("couldn't parse timestamp %s, invalid format", in)
+	}
+	timestampComponents := strings.Split(in, ":")
+	if len(timestampComponents) != 2 {
+		return 0, fmt.Errorf("invalid timestamp format: %s", in)
+	}
+	timeStampHours, err := strconv.Atoi(timestampComponents[0])
+	if err != nil {
+		return 0, err
+	}
+	timeStampMinutes, err := strconv.Atoi(timestampComponents[1])
+	if err != nil {
+		return 0, err
+	}
+	if timeStampHours < 0 || timeStampHours > 24 || timeStampMinutes < 0 || timeStampMinutes > 60 {
+		return 0, fmt.Errorf("timestamp %s out of range", in)
+	}
+	// Timestamps are stored as minutes elapsed in the day, so multiply hours by 60
+	mins = timeStampHours*60 + timeStampMinutes
+	return mins, nil
+}
+
+// Converts a range that can be represented as strings (e.g. monday:wednesday) into an equivalent integer-represented range
+func stringableRangeFromString(in string, r stringableRange) (err error) {
+	in = strings.ToLower(in)
+	if strings.ContainsRune(in, ':') {
+		components := strings.Split(in, ":")
+		if len(components) != 2 {
+			return fmt.Errorf("couldn't parse range %s, invalid format", in)
+		}
+		start, err := r.memberFromString(components[0])
+		if err != nil {
+			return err
+		}
+		End, err := r.memberFromString(components[1])
+		if err != nil {
+			return err
+		}
+		r.setBegin(start)
+		r.setEnd(End)
+		return nil
+	}
+	val, err := r.memberFromString(in)
+	if err != nil {
+		return err
+	}
+	r.setBegin(val)
```

Comment:
Do we set them individualy somewhere? Can we do with one setter function ? Do we need those extra setters, can't we set EndDate directly?